Sanity test

chuanchang opened the pull request:
Add sanity test for atomic commands{info,install,images,mount,stop,uninstall,unmount,update,version} testing, and removing previous seperated feature, which has been covered by sanity test now.

```
bash-4.3# behave  features/rhelah/atomic/sanity_test.feature
@atomic @host_subscribed @ah_upgrade
Feature: Atomic host sanity test # features/rhelah/atomic/sanity_test.feature:2
  Describes the basic 'atomic host' command testRegister all hosts to stage

  Background: Atomic hosts are discovered  # features/rhelah/atomic/sanity_test.feature:5

  Scenario: 1. Pull latest image from repository       # features/rhelah/atomic/sanity_test.feature:8
    Given "all" hosts can be pinged                    # steps/common.py:201 4.064s
    Given List all locally installed container images  # steps/rhelah.py:532 5.216s
    When atomic update latest "centos" from repository # steps/rhelah.py:498 16.512s
    Then Check whether "centos" is installed           # steps/rhelah.py:507 4.806s

  Scenario: 2. mount image to a specified directory         # features/rhelah/atomic/sanity_test.feature:13
    Given "all" hosts can be pinged                         # steps/common.py:201 4.305s
    When atomic mount "image" to a specified "/mnt"         # steps/rhelah.py:455 14.432s
    Then check whether atomic mount point "/var/mnt" exists # steps/rhelah.py:475 6.684s

  Scenario: 3. unmount image previously mounted                     # features/rhelah/atomic/sanity_test.feature:17
    Given "all" hosts can be pinged                                 # steps/common.py:201 5.839s
    When atomic unmount image from previous "/mnt"                  # steps/rhelah.py:482 5.772s
    Then check whether atomic mount point "/var/mnt" does not exist # steps/rhelah.py:492 4.062s

  Scenario: 4. docker run busybox with detach mode     # features/rhelah/atomic/sanity_test.feature:21
    Given "all" hosts can be pinged                    # steps/common.py:201 4.014s
    When docker run "centos" detach mode with "top -b" # steps/docker.py:64 4.465s
    Then check whether there is a running container    # steps/docker.py:70 4.073s

  Scenario: 5. mount running container to a specified directory  # features/rhelah/atomic/sanity_test.feature:25
    Given "all" hosts can be pinged                              # steps/common.py:201 4.058s
    When atomic mount "container" to a specified "/mnt"          # steps/rhelah.py:455 9.842s
    Then check whether atomic mount point "/var/mnt" exists      # steps/rhelah.py:475 4.699s

  Scenario: 6. unmount container previously mounted                 # features/rhelah/atomic/sanity_test.feature:29
    Given "all" hosts can be pinged                                 # steps/common.py:201 4.041s
    When atomic unmount container from previous "/mnt"              # steps/rhelah.py:482 4.677s
    Then check whether atomic mount point "/var/mnt" does not exist # steps/rhelah.py:492 4.018s

  Scenario: 7. atomic stop previous running container  # features/rhelah/atomic/sanity_test.feature:33
    Given "all" hosts can be pinged                    # steps/common.py:201 4.053s
    When atomic stop container                         # steps/rhelah.py:446 8.656s

  Scenario: 8. Build a new image from Dockerfile                                                                                   # features/rhelah/atomic/sanity_test.feature:36
    Given "all" hosts can be pinged                                                                                                # steps/common.py:201 4.139s
    When docker build an image from "https://raw.githubusercontent.com/projectatomic/atomic/master/tests/test-images/Dockerfile.1" # steps/docker.py:76 9.300s
    Then Check whether dangling images exist                                                                                       # steps/rhelah.py:539 4.663s

  Scenario: 9. Remove dangling images               # features/rhelah/atomic/sanity_test.feature:40
    Given "all" hosts can be pinged                 # steps/common.py:201 4.025s
    When Remove all dangling images                 # steps/rhelah.py:548 4.660s
    Then Check whether dangling images do not exist # steps/rhelah.py:555 4.804s

  Scenario: 10. Build a new image from Dockerfile                                                                                                          # features/rhelah/atomic/sanity_test.feature:44
    Given "all" hosts can be pinged                                                                                                                        # steps/common.py:201 4.039s
    When docker build an image with tag "centos_label" from "https://raw.githubusercontent.com/projectatomic/atomic/master/tests/test-images/Dockerfile.1" # steps/docker.py:76 8.420s
    Then Check whether "centos_label" is installed                                                                                                         # steps/rhelah.py:507 4.481s

  Scenario: 11. Display image label of name version release             # features/rhelah/atomic/sanity_test.feature:48
    Given "all" hosts can be pinged                                     # steps/common.py:201 4.010s
    When Display "centos_label" "atomic-test-1" of name version release # steps/rhelah.py:618 4.517s

  Scenario: 12. Read the LABEL INSTALL field in the container image                             # features/rhelah/atomic/sanity_test.feature:51
    Given "all" hosts can be pinged                                                             # steps/common.py:201 3.981s
    When Read the LABEL INSTALL "I am the install label." field in the container "centos_label" # steps/rhelah.py:639 5.954s

  Scenario: 13. Build a new image with new tag from Dockerfile                                                                                             # features/rhelah/atomic/sanity_test.feature:54
    Given "all" hosts can be pinged                                                                                                                        # steps/common.py:201 6.338s
    When docker build an image with tag "scratch_test" from "https://raw.githubusercontent.com/projectatomic/atomic/master/tests/test-images/Dockerfile.2" # steps/docker.py:76 5.013s
    Then Check whether "scratch_test" is installed                                                                                                         # steps/rhelah.py:507 9.247s

  Scenario: 14. Display label information about a local image       # features/rhelah/atomic/sanity_test.feature:58
    Given "all" hosts can be pinged                                 # steps/common.py:201 4.004s
    When Display LABEL information about an image "scratch_test"    # steps/rhelah.py:602 4.820s
    Then Check LABEL "Name: atomic-test-2" information for an image # steps/rhelah.py:611 0.000s

  Scenario: 15. Display label information about a remote image        # features/rhelah/atomic/sanity_test.feature:62
    Given "all" hosts can be pinged                                   # steps/common.py:201 4.083s
    When Display LABEL information about a "remote" image "rhel7"     # steps/rhelah.py:602 6.856s
    Then Check LABEL "Vendor: Red Hat, Inc." information for an image # steps/rhelah.py:611 0.000s

  Scenario: 16. Remove all of containers  # features/rhelah/atomic/sanity_test.feature:66
    Given "all" hosts can be pinged       # steps/common.py:201 4.122s
    When docker remove all of containers  # steps/docker.py:92 4.315s

  Scenario: 17. Remove scratch_test image                    # features/rhelah/atomic/sanity_test.feature:69
    Given "all" hosts can be pinged                          # steps/common.py:201 4.139s
    When Remove "scratch_test" from system                   # steps/rhelah.py:516 5.027s
    Then Check whether "scratch_test" is removed from system # steps/rhelah.py:523 4.575s

  Scenario: 18. Remove centos_label image                    # features/rhelah/atomic/sanity_test.feature:73
    Given "all" hosts can be pinged                          # steps/common.py:201 4.103s
    When Remove "centos_label" from system                   # steps/rhelah.py:516 4.686s
    Then Check whether "centos_label" is removed from system # steps/rhelah.py:523 4.960s

  Scenario: 19. Remove centos image                   # features/rhelah/atomic/sanity_test.feature:77
    Given "all" hosts can be pinged                   # steps/common.py:201 4.386s
    Then remove docker image "centos"                 # steps/docker.py:44 4.223s
    And Check whether "centos" is removed from system # steps/rhelah.py:523 4.686s

  Scenario: 20. Rollback to the original deployment                              # features/rhelah/atomic/sanity_test.feature:81
    Given "all" hosts can be pinged                                              # steps/common.py:201 4.108s
    Given there is "2" atomic host tree deployed                                 # steps/rhelah.py:224 4.478s
    And the original atomic version has been recorded                            # steps/rhelah.py:282 4.483s
    When atomic host rollback is successful                                      # steps/rhelah.py:341 6.313s
    And wait "60" seconds for "all" to reboot                                    # steps/rhelah.py:143 75.458s
    Then the current atomic version should not match the original atomic version # steps/rhelah.py:330 4.221s
Unregister all hosts

1 feature passed, 0 failed, 0 skipped
20 scenarios passed, 0 failed, 0 skipped
60 steps passed, 0 failed, 0 skipped, 0 undefined
Took 6m23.923s
```

jlebon commented:
I'm not sure how I feel about putting them all together in one place. I do like the additional coverage you added (e.g. checking the INSTALL label), but to me personally, it's nicer to have separated, smaller, focused feature tests for different functionalities of `/bin/atomic`.

Smaller feature tests would make them more legible and maintainable and would allow for more easily cherry-picking tests when debugging something. Also, in a large feature test, it's hard to keep track of which step depends on the state created by which previous step.

I'll let others chime in with their thoughts about this.

chuanchang commented:
@jlebon, thanks for your comments. I will try to answer your question then raise my question.

> Smaller feature tests would make them more legible and maintainable and would allow for more easily cherry-picking tests when debugging something.

Yes, but we can also do this in bigger test feature to use tag, it looks like this, if it's what you want, I will do it.

```
  @build_label_image
  Scenario: 10. Build a new image from Dockerfile
       When docker build an image with tag "centos_label" from "https://raw.githubusercontent.com/projectatomic/atomic/master/tests/test-images/Dockerfile.1"
       Then Check whether "centos_label" is installed

  @display_image_label_of_name_version
  Scenario: 11. Display image label of name version release
       When Display "centos_label" "atomic-test-1" of name version release

  @read_image_label_install
  Scenario: 12. Read the LABEL INSTALL field in the container image
       When Read the LABEL INSTALL "I am the install label." field in the container "centos_label"
```

If you only want to run 'LABEL INSTALL' test, your behave cmdline looks like this.

``` shell
# behave -t @build_label_image, @read_image_label_install features/rhelah/atomic/sanity_test.feature
```

> in a large feature test, it's hard to keep track of which step depends on the state created by which previous step.

Yes, it's indeed a question, but at present, the registration, upgrading and pulling images will kill us much more time, because smaller feature tests will do them each time, in fact, we may resue previous test environment, you see we only Took 6m23.923s in above sanity tests, or I may reserve previous existing smaller features tests for seperated feature tests if you need.

miabbott commented:
I agree with @jlebon that I would prefer that the features be smaller.  But rather than break out the tests in to smaller chunks, I think the approach that @chuanchang suggested by using tags would be a good compromise. 

Another point is that even though the sanity feature test has 20 scenarios defined, each one of those can be reused in a separate test, so its not as if all the scenarios are pinned to that single test.

My vote is to keep the sanity test as is (since the work has already been done), but tag the scenarios effectively so that individual scenarios can be cherry-picked for more easy debugging (as @jlebon suggested).

jlebon commented:
Hmm, yes I suppose tags will do the trick. Maybe prefix them all with e.g. `atomic_sanity_test`? (Or something shorter if you like).

chuanchang commented:
@jlebon, a prefix for all of tags may be not necessary, because sanity_test.feature under the features/rhelah/atomic directory, which is only used for putting atomic cmdline related test scrips now.

jlebon commented on features/rhelah/atomic/sanity_test.feature:
```diff
@@ -0,0 +1,106 @@
+
+  @build_dangling_image
+  Scenario: 8. Build a new image from Dockerfile
+       When docker build an image from "https://raw.githubusercontent.com/projectatomic/atomic/master/tests/test-images/Dockerfile.1"
```
I know this is carry-over from the previously split feature tests, but I'm wondering if we should instead keep a copy of this Dockerfile locally (or just directly create our own). That way, we're not tied to whatever changes might occur to it there, and it also gives us more flexibility to have a Dockerfile with everything we need for testing. Up to you, just thinking out loud here. :)

chuanchang replied:
@jlebon, yes, we have the same idea, ypu and I had ever considered to create a resources directory then put test related test files like Dockerfile.*, *.yaml, etc into it, I will make it happen in next patches.

jlebon commented on steps/docker.py:
```diff
@@ -82,3 +82,15 @@ def step_impl(context, dockerfile, tag=''):
+def step_impl(context):
+   '''Remove all of images'''
+   assert context.remote_cmd('shell',
+                             module_args='docker rmi `docker images -a -q`')
```
Won't this fail if there are no images present? Maybe use `xargs -r` instead? E.g. `docker images -a -q | xargs -r docker rmi`. The `-r` is nifty:

```
 -r, --no-run-if-empty
        If  the standard input does not contain any nonblanks, do not run the command.  Normally, the
        command is run once even if there is no input.  This option is a GNU extension.
```

chuanchang replied:
Good trick.

jlebon commented on steps/docker.py:
```diff
@@ -82,3 +82,15 @@ def step_impl(context, dockerfile, tag=''):
+def step_impl(context):
+   '''Remove all of containers'''
+   assert context.remote_cmd('shell',
+                             module_args='docker rm `docker ps -a -q`')
```
Same comment as above.

chuanchang replied:
@jlebon, thanks for your good advise :)

chuanchangjia committed:
steps.rhelah: update rhelah module

In order to avoid mounting context changes in /proc/mounts, only check
mount point in find_mount_point() instead, and added three new test
libraied into rhelah module.

Signed-off-by: Alex Jia <ajia@redhat.com>

chuanchangjia committed:
steps.docker: update docker module

Using local dockerfile instead of a URL, and added two new test
libraried into docker module.

Signed-off-by: Alex Jia <ajia@redhat.com>

chuanchangjia committed:
Refactoring test directory

Add new resources/{docker,kube} directory and move previous
steps/kube_resources/* to resources/kube, and added new
dockerfiles related file into resources/docker, meanwhile,
to update steps/kubernetes.py to follow this change and
remove steps/kube_resources directory.

Signed-off-by: Alex Jia <ajia@redhat.com>

chuanchangjia committed:
features.rhelah.atomic: add sanity test feature for atomic commands test

Add sanity test feature to cover atomic commands test and remove prevous
existing small features test.

Signed-off-by: Alex Jia <ajia@redhat.com>

chuanchang commented:
@jlebon, it seems it's not trivial work to meet your all of requirements :)

chuanchang commented:
The following is new test result, as you seen, it exists a known bug 1273365.

```
bash-4.3# behave features/rhelah/atomic/sanity_test.feature 
@atomic @host_subscribed @ah_upgrade
Feature: Atomic host sanity test # features/rhelah/atomic/sanity_test.feature:2
  Describes the basic 'atomic host' command testRegister all hosts to stage

  Background: Atomic hosts are discovered  # features/rhelah/atomic/sanity_test.feature:5

  @pull_image
  Scenario: 1. Pull latest image from repository       # features/rhelah/atomic/sanity_test.feature:9
    Given "all" hosts can be pinged                    # steps/common.py:201 7.835s
    Given List all locally installed container images  # steps/rhelah.py:532 9.254s
    When atomic update latest "centos" from repository # steps/rhelah.py:498 20.068s
    Then Check whether "centos" is installed           # steps/rhelah.py:507 8.281s

  @mount_image_to_path
  Scenario: 2. mount image to a specified directory         # features/rhelah/atomic/sanity_test.feature:15
    Given "all" hosts can be pinged                         # steps/common.py:201 7.034s
    When atomic mount "image" to a specified "/mnt"         # steps/rhelah.py:455 16.760s
    Then check whether atomic mount point "/var/mnt" exists # steps/rhelah.py:475 7.545s

  @unmount_image_from_path
  Scenario: 3. unmount image previously mounted                     # features/rhelah/atomic/sanity_test.feature:20
    Given "all" hosts can be pinged                                 # steps/common.py:201 7.445s
    When atomic unmount image from previous "/mnt"                  # steps/rhelah.py:482 8.591s
      Assertion Failed: Error while running 'atomic unmount'

Ansible Output: {'dark': {}, 'contacted': {u'10.8.48.220': {'cmd': ['atomic', 'unmount', '/mnt'], 'end': '2015-10-20 10:07:34.822040', 'stdout': u'', 'changed': True, 'start': '2015-10-20 10:07:34.001632', 'delta': '0:00:00.820408', 'stderr': 'Device mounted at /mnt is not a docker container.', 'rc': 1, 'invocation': {'module_name': 'command', 'module_args': u'atomic unmount /mnt'}, 'warnings': []}}}
    Then check whether atomic mount point "/var/mnt" does not exist # None

  @run_container_in_bg
  Scenario: 4. docker run busybox with detach mode     # features/rhelah/atomic/sanity_test.feature:25
    Given "all" hosts can be pinged                    # steps/common.py:201 7.713s
    When docker run "centos" detach mode with "top -b" # steps/docker.py:65 7.830s
    Then check whether there is a running container    # steps/docker.py:71 7.597s

  @mount_container_to_path
  Scenario: 5. mount running container to a specified directory  # features/rhelah/atomic/sanity_test.feature:30
    Given "all" hosts can be pinged                              # steps/common.py:201 8.596s
    When atomic mount "container" to a specified "/mnt"          # steps/rhelah.py:455 17.082s
    Then check whether atomic mount point "/var/mnt" exists      # steps/rhelah.py:475 7.644s

  @unmount_container_from_path
  Scenario: 6. unmount container previously mounted                 # features/rhelah/atomic/sanity_test.feature:35
    Given "all" hosts can be pinged                                 # steps/common.py:201 7.841s
    When atomic unmount container from previous "/mnt"              # steps/rhelah.py:482 9.186s
      Assertion Failed: Error while running 'atomic unmount'

Ansible Output: {'dark': {}, 'contacted': {u'10.8.48.220': {'cmd': ['atomic', 'unmount', '/mnt'], 'end': '2015-10-20 10:08:47.714500', 'stdout': u'', 'changed': True, 'start': '2015-10-20 10:08:46.657548', 'delta': '0:00:01.056952', 'stderr': 'Device mounted at /mnt is not a docker container.', 'rc': 1, 'invocation': {'module_name': 'command', 'module_args': u'atomic unmount /mnt'}, 'warnings': []}}}
    Then check whether atomic mount point "/var/mnt" does not exist # None

  @stop_container
  Scenario: 7. atomic stop previous running container  # features/rhelah/atomic/sanity_test.feature:40
    Given "all" hosts can be pinged                    # steps/common.py:201 8.767s
    When atomic stop container                         # steps/rhelah.py:446 16.619s

  @build_dangling_image
  Scenario: 8. Build a new image from Dockerfile         # features/rhelah/atomic/sanity_test.feature:44
    Given "all" hosts can be pinged                      # steps/common.py:201 11.702s
    When docker build an image from local "Dockerfile.1" # steps/docker.py:77
/usr/lib/python2.7/site-packages/behave/runner.py:290: ContextMaskWarning: behave runner is masking context attribute 'table' orignally set in runner.context.table = self.table (/usr/lib/python2.7/site-packages/behave/model.py:1455)
  self.table = original_table
/usr/lib/python2.7/site-packages/behave/runner.py:291: ContextMaskWarning: behave runner is masking context attribute 'text' orignally set in runner.context.text = self.text (/usr/lib/python2.7/site-packages/behave/model.py:1454)
  self.text  = original_text
/root/UATFramework/environment.py:145: ContextMaskWarning: behave runner is masking context attribute 'result' orignally set in **kwargs (environment.py:145)
    When docker build an image from local "Dockerfile.1" # steps/docker.py:77 34.465s
    Then Check whether dangling images exist             # steps/rhelah.py:539 8.764s

  @remove_dangling_image
  Scenario: 9. Remove dangling images               # features/rhelah/atomic/sanity_test.feature:49
    Given "all" hosts can be pinged                 # steps/common.py:201 8.151s
    When Remove all dangling images                 # steps/rhelah.py:548 8.459s
    Then Check whether dangling images do not exist # steps/rhelah.py:555 9.228s

  @build_label_image
  Scenario: 10. Build a new image from Dockerfile                                # features/rhelah/atomic/sanity_test.feature:54
    Given "all" hosts can be pinged                                              # steps/common.py:201 7.974s
    When docker build an image with tag "centos_label" from local "Dockerfile.1" # steps/docker.py:77 31.926s
    Then Check whether "centos_label" is installed                               # steps/rhelah.py:507 8.414s

  @check_image_name_version
  Scenario: 11. Display image label of name version release             # features/rhelah/atomic/sanity_test.feature:59
    Given "all" hosts can be pinged                                     # steps/common.py:201 8.074s
    When Display "centos_label" "atomic-test-1" of name version release # steps/rhelah.py:618 8.762s

  @read_label_install_in_image
  Scenario: 12. Read the LABEL INSTALL field in the container image                             # features/rhelah/atomic/sanity_test.feature:63
    Given "all" hosts can be pinged                                                             # steps/common.py:201 7.544s
    When Read the LABEL INSTALL "I am the install label." field in the container "centos_label" # steps/rhelah.py:639 8.826s

  @build_info_image
  Scenario: 13. Build a new image with new tag from Dockerfile                   # features/rhelah/atomic/sanity_test.feature:67
    Given "all" hosts can be pinged                                              # steps/common.py:201 7.422s
    When docker build an image with tag "scratch_test" from local "Dockerfile.2" # steps/docker.py:77 30.811s
    Then Check whether "scratch_test" is installed                               # steps/rhelah.py:507 8.552s

  @check_local_image_label
  Scenario: 14. Display label information about a local image       # features/rhelah/atomic/sanity_test.feature:72
    Given "all" hosts can be pinged                                 # steps/common.py:201 7.990s
    When Display LABEL information about an image "scratch_test"    # steps/rhelah.py:602 9.277s
    Then Check LABEL "Name: atomic-test-2" information for an image # steps/rhelah.py:611 0.000s

  @check_remote_image_label
  Scenario: 15. Display label information about a remote image        # features/rhelah/atomic/sanity_test.feature:77
    Given "all" hosts can be pinged                                   # steps/common.py:201 7.026s
    When Display LABEL information about a "remote" image "rhel7"     # steps/rhelah.py:602 10.982s
    Then Check LABEL "Vendor: Red Hat, Inc." information for an image # steps/rhelah.py:611 0.000s

  @remove_built_info_image
  Scenario: 16. Remove scratch_test image                    # features/rhelah/atomic/sanity_test.feature:82
    Given "all" hosts can be pinged                          # steps/common.py:201 8.081s
    When Remove "scratch_test" from system                   # steps/rhelah.py:516 8.413s
    Then Check whether "scratch_test" is removed from system # steps/rhelah.py:523 9.582s

  @remove_built_label_image
  Scenario: 17. Remove centos_label image                    # features/rhelah/atomic/sanity_test.feature:87
    Given "all" hosts can be pinged                          # steps/common.py:201 7.309s
    When Remove "centos_label" from system                   # steps/rhelah.py:516 8.869s
    Then Check whether "centos_label" is removed from system # steps/rhelah.py:523 9.209s

  @remove_pulled_image
  Scenario: 18. Remove centos image                   # features/rhelah/atomic/sanity_test.feature:92
    Given "all" hosts can be pinged                   # steps/common.py:201 7.562s
    Then remove docker image "centos"                 # steps/docker.py:45 8.584s
    And Check whether "centos" is removed from system # steps/rhelah.py:523 8.597s

  @rollback_host
  Scenario: 19. Rollback to the original deployment                              # features/rhelah/atomic/sanity_test.feature:97
    Given "all" hosts can be pinged                                              # steps/common.py:201 8.026s
    Given there is "2" atomic host tree deployed                                 # steps/rhelah.py:224 9.588s
    And the original atomic version has been recorded                            # steps/rhelah.py:282 8.781s
    When atomic host rollback is successful                                      # steps/rhelah.py:341 12.171s
    And wait "60" seconds for "all" to reboot                                    # steps/rhelah.py:143 89.496s
    Then the current atomic version should not match the original atomic version # steps/rhelah.py:330 7.467s
Unregister all hosts


Failing scenarios:
  features/rhelah/atomic/sanity_test.feature:20  3. unmount image previously mounted
  features/rhelah/atomic/sanity_test.feature:35  6. unmount container previously mounted

0 features passed, 1 failed, 0 skipped
17 scenarios passed, 2 failed, 0 skipped
54 steps passed, 2 failed, 2 skipped, 0 undefined
Took 10m47.770s 
```

jlebon commented on steps/rhelah.py:
```diff
@@ -613,3 +613,38 @@ def step_impl(context, label):
+    '''Display image label of name version release'''
+    image_info = context.remote_cmd(cmd='command',
+                                    module_args='atomic version %s' % image)
+    if not label:
```
Shouldn't this be `if not image_info` ?

chuanchang replied:
Nope, if the label is empty string then assert image_info.

jlebon commented on steps/rhelah.py:
```diff
@@ -613,3 +613,38 @@ def step_impl(context, label):
+    else:
+        assert label in image_info[0]['stdout'], \
+            ("The current LABEL INSTALL %s " % image_info +
+             "does not match the expected LABEL INSTALL %s" % label)
```
Is that the correct error message? Maybe something like:

``` python
   "The LABEL %s in image %s " % (label, image) +
   "does not match the expected LABEL value %s" % label
```

chuanchang replied:
Hmmm, it's not a big deal, I will update it :)

jlebon commented on steps/rhelah.py:
```diff
@@ -613,3 +613,38 @@ def step_impl(context, label):
+    '''Check LABEL INSTALL in the image'''
+    install_label = context.remote_cmd(cmd='command',
+                                       module_args='atomic install %s' % image)
+    if not label:
```
Here too, shouldn't that be `if not install_label` ?

chuanchang replied:
same as above.

jlebon commented:
Looking good! We're almost there. :)

Just made a few final comments. Also, it looks like the `install.sh` and `Dockerfile.3` files are not needed, so we should probably get rid of them.

And good catch on the mount/unmount bug!

chuanchang commented:
@jlebon, okay, I will remove the install.sh and Dockerfile.3 files from this PR then add them in the next PR, thanks for your comments again.

jlebon commented:
LGTM! :+1:

miabbott commented:
It looks like all the comments have been addressed.  Thanks to @jlebon for the thorough review.  And thanks for @chuanchang for working through the process.
